src/System.Management.Automation/engine/MshObject.cs

```diff
@@ -2347,6 +2378,204 @@ internal bool IsHelpObject
+
+        /// <summary>
+        /// Initializes a new instance of PSObjectBuilder.
+        /// </summary>
+        /// <param name="validatedNameCount">The initial size of the set that tracks validated parameter names.</param>
+        public PSObjectBuilder(int validatedNameCount)
+        {
+            if (validatedNameCount < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(validatedNameCount));
+
+            }
+            _validatedNames = new HashSet<string>(validatedNameCount);
+            _currentObjectStack = null;
+            _currentObject = null;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of PSObjectBuilder.
+        /// </summary>
+        /// <param name="validatedNameCount">The initial size of the set that tracks validated parameter names.</param>
+        /// <param name="trustedNames">True if all the names are preValidated.</param>
+        internal PSObjectBuilder(int validatedNameCount, bool trustedNames)
+        {
+            if (validatedNameCount < 1)
+            {
+                throw new ArgumentOutOfRangeException(nameof(validatedNameCount));
+
+            }
+
+            _validatedNames = null;
+            _currentObjectStack = null;
+            _currentObject = null;
+        }
+
+        /// <summary>
+        /// Begins the creation of a new PSObject
+        /// </summary>
+        /// <param name="capacity">The initial capacity for the members collection.</param>
+        public void BeginCreateObject(int capacity)
+        {
+            if (_currentObject != null)
+            {
+                CurrentObjectStack.Push(_currentObject);
+            }
+            _currentObject = new PSObject(capacity);
+        }
+
+        /// <summary>
+        /// Begins the creation of a new PSObject
+        /// </summary>
+        /// <param name="instance">The object to wrap in a PSObject.</param>
+        /// <param name="capacity">The initial capacity for the members collection.</param>
+        public void BeginCreateObject(object instance, int capacity)
+        {
+            if (_currentObject != null)
+            {
+                CurrentObjectStack.Push(_currentObject);
+            }
+            _currentObject = new PSObject(instance, capacity);
+        }
+
+
+        /// <summary>
+        /// Add a property to the object.
+        /// </summary>
+        /// <param name="name">The name of the property.</param>
+        /// <param name="value">The value of the property.</param>
+        public void AddNoteProperty(string name, object value)
+        {
+            var validated = _validatedNames?.Contains(name) ?? true;
+            if (!validated)
+            {
+                _validatedNames.Add(name);
+            }
+            _currentObject.Properties.Add(new PSNoteProperty(name, value), validated, inObjectBuilderScope:true);
```

Comment:
Seem we should check  `_currentObject != null`.

For discussion I should mention that we could use `UnsafeAdd()` where internally inObjectBuilderScope:true.